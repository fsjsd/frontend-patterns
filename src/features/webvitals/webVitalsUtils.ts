/* eslint-disable @typescript-eslint/no-explicit-any */
/* istanbul ignore file */

import {
  FirstInputPolyfillEntry,
  getCLS,
  getFCP,
  getFID,
  getLCP,
  getTTFB,
  Metric,
  NavigationTimingPolyfillEntry,
} from 'web-vitals'

// adapted from:
// https://github.com/GoogleChromeLabs/web-vitals-report/blob/main/src/js/analytics.js

type WebVitalPerfEntries = (
  | PerformanceEntry
  | FirstInputPolyfillEntry
  | NavigationTimingPolyfillEntry
)[]

interface RegisterWebVitalsListenersProps {
  onCls: (m: MetricInfo) => void
  onFcp: (m: MetricInfo) => void
  onFid: (m: MetricInfo) => void
  onLcp: (m: MetricInfo) => void
  onTtfb: (m: MetricInfo) => void
}

export interface MetricInfo {
  value: number
  event_category: string
  event_label: string
  metric_value: number
  metric_delta: number
  metric_rating: string
  non_interaction: boolean
  debug_target?: string
  debug_event?: string
  debug_timing?: string
  fetch_start?: number
  domain_lookup_start?: number
  connect_start?: number
  request_start?: number
  response_start?: number
}

const thresholds: Record<string, any> = {
  CLS: [0.1, 0.25],
  FCP: [1800, 3000],
  FID: [100, 300],
  LCP: [2500, 4000],
  TTFB: [800, 1800],
}

function getRating(value: number, thresholdTuple: number[]) {
  if (value > thresholdTuple[1]) {
    return 'poor'
  }
  if (value > thresholdTuple[0]) {
    return 'ni'
  }
  return 'good'
}

function getSelector(node: Element, maxLen = 100) {
  let sel = ''
  try {
    while (node && node.nodeType !== 9) {
      const part = node.id
        ? '#' + node.id
        : node.nodeName.toLowerCase() +
          (node.className && node.className.length
            ? '.' + Array.from(node.classList.values()).join('.')
            : '')
      if (sel.length + part.length > maxLen - 1) return sel || part
      sel = sel ? part + '>' + sel : part
      if (node.id) break
      node = node.parentNode as any
    }
  } catch (err) {
    // Do nothing...
  }
  return sel
}

function getLargestLayoutShiftEntry(entries: any) {
  return entries.reduce((a: any, b: any) => (a && a.value > b.value ? a : b))
}

function getLargestLayoutShiftSource(sources: any) {
  return sources.reduce((a: any, b: any) => {
    return a.node &&
      a.previousRect.width * a.previousRect.height >
        b.previousRect.width * b.previousRect.height
      ? a
      : b
  })
}

function wasFIDBeforeDCL(fidEntry: FirstInputPolyfillEntry) {
  const navEntry = performance.getEntriesByType('navigation')[0] as any
  return navEntry && fidEntry.startTime < navEntry.domContentLoadedEventStart
}

function getDebugInfo(name: string, entries: WebVitalPerfEntries = []) {
  // In some cases there won't be any entries (e.g. if CLS is 0,
  // or for LCP after a bfcache restore), so we have to check first.
  if (entries.length) {
    if (name === 'LCP') {
      const lastEntry = entries[entries.length - 1] as any
      return {
        debug_target: getSelector(lastEntry.element),
        event_time: lastEntry.startTime,
      }
    } else if (name === 'FID') {
      const firstEntry = entries[0] as FirstInputPolyfillEntry
      return {
        debug_target: firstEntry.target ? getSelector(firstEntry.target) : '',
        debug_event: firstEntry.name,
        debug_timing: wasFIDBeforeDCL(firstEntry) ? 'pre_dcl' : 'post_dcl',
        event_time: firstEntry.startTime,
      }
    } else if (name === 'CLS') {
      const largestEntry = getLargestLayoutShiftEntry(entries)
      if (largestEntry && largestEntry.sources && largestEntry.sources.length) {
        const largestSource = getLargestLayoutShiftSource(largestEntry.sources)
        if (largestSource) {
          return {
            debug_target: getSelector(largestSource.node),
            event_time: largestEntry.startTime,
          }
        }
      }
    }
  }
  // Return default/empty params in case there are no entries.
  return {
    debug_target: '(not set)',
  }
}

const mapMetric = (metric: Metric) => {
  const { name, value, delta, id, entries } = metric
  const params: MetricInfo = {
    value: Math.round(name === 'CLS' ? delta * 1000 : delta),
    event_category: 'Web Vitals',
    event_label: id,
    metric_value: value,
    metric_delta: delta,
    metric_rating: getRating(value, thresholds[name]),
    non_interaction: true,
    ...getDebugInfo(name, entries),
  }

  if (name === 'TTFB' && entries.length) {
    const navEntry = entries[0] as any
    Object.assign(params, {
      fetch_start: navEntry.fetchStart,
      domain_lookup_start: navEntry.domainLookupStart,
      connect_start: navEntry.connectStart,
      request_start: navEntry.requestStart,
      response_start: navEntry.responseStart,
    })
  }

  console.log(`[${name}]`, params)

  return params
}

export const registerWebVitalsListeners = ({
  onCls,
  onFcp,
  onLcp,
  onTtfb,
  onFid,
}: RegisterWebVitalsListenersProps) => {
  getCLS(m => onCls(mapMetric(m)))
  getFCP(m => onFcp(mapMetric(m)))
  getFID(m => onFid(mapMetric(m)))
  getLCP(m => onLcp(mapMetric(m)))
  getTTFB(m => onTtfb(mapMetric(m)))
}
/*
const getWebVitalAsync = async (getWebVitalFn: (onReport: ReportHandler, reportAllChanges?: boolean | undefined) => void) => {
  return new Promise((resolve, reject) => {
    getWebVitalFn((metric) => {
      resolve(handleMetric(metric));
    })  
  });
};

export async function cls() {
  const result = await getWebVitalAsync(getCLS);
  return result;
}

export async function fcp() {
  const result = await getWebVitalAsync(getFCP);
  return result;
}

export async function fid() {
  const result = await getWebVitalAsync(getFID);
  return result;
}

export async function lcp() {
  const result = await getWebVitalAsync(getLCP);
  return result;
}

export async function ttfb() {
  const result = await getWebVitalAsync(getTTFB);
  return result;
}
*/
