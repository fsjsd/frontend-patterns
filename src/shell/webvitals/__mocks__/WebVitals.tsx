import React from 'react'

const WebVitals = () => {
  return (
    <div>WebVitals Mock</div>
  )
}

export default WebVitals