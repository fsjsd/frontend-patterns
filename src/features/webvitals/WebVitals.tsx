import React, { useEffect, useState } from 'react'
import styled from 'styled-components'
import ToolTip from '../../ux/ToolTip'
import { MetricInfo, registerWebVitalsListeners } from './webVitalsUtils'

const ClsWrapper = styled.div`
  display: flex;
  flex-direction: row;
  > div {
    margin-right: 10px;
    display: flex;
    align-items: center;
    line-height:normal;
  }
`

const MetricRating = styled.div<{ rating?: string }>`
  display: ${props => (props.rating ? 'inline-block' : 'none')};
  color: white;
  border-radius: 10px;
  padding: 7px;
  margin: 0 3px;
  background-color: ${props => {
    switch (props.rating) {
      case 'good':
        return '#2daf2d'
      case 'poor':
        return '#972020'
      case 'ni':
        return '#a86323'
      default:
        return '#aaaaaa'
    }
  }};
`

const mapMetricRating = (rating: string) => {
  switch (rating) {
    case 'good':
      return 'Good'
    case 'poor':
      return 'Poor'
    case 'ni':
      return 'Needs improvement'
    default:
      return 'unknown'
  }
}

// https://web.dev/fid/
// https://web.dev/ttfb/
// https://web.dev/fcp/
// https://web.dev/cls/

const WebVitals = () => {
  const [wvCls, setWvCls] = useState<MetricInfo>()
  const [wvFcp, setWvFcp] = useState<MetricInfo>()
  const [wvFid, setWvFid] = useState<MetricInfo>()
  const [wvLcp, setWvLcp] = useState<MetricInfo>()
  const [wvTtfb, setWvTtfb] = useState<MetricInfo>()

  const [clsTooltipVisible, setClsTooltipVisible] = useState<boolean>(false)
  const [fcpTooltipVisible, setFcpTooltipVisible] = useState<boolean>(false)
  const [fidTooltipVisible, setFidTooltipVisible] = useState<boolean>(false)
  const [lcpTooltipVisible, setLcpTooltipVisible] = useState<boolean>(false)
  const [ttfbTooltipVisible, setTtfbTooltipVisible] = useState<boolean>(false)


  useEffect(() => {
    registerWebVitalsListeners({
      onCls: setWvCls,
      onFcp: setWvFcp,
      onFid: setWvFid,
      onLcp: setWvLcp,
      onTtfb: setWvTtfb,
    })
  }, [])

  return (
    <ClsWrapper>
      {wvLcp && (
        <div
          aria-label='LCP'
          title="Largest Contentful Paint"
          onMouseEnter={() => setLcpTooltipVisible(true)}
          onMouseLeave={() => setLcpTooltipVisible(false)}
        >
          <ToolTip isVisible={lcpTooltipVisible}>
            <b>Largest Contentful Paint.</b>
            <br />
            {mapMetricRating(wvLcp.metric_rating)}. {wvLcp.value / 1000}s
          </ToolTip>
          LCP: <MetricRating rating={wvLcp.metric_rating} />
        </div>
      )}
      {wvFid && (
        <div
          title="First Input Delay"
          aria-label='FID'
          onMouseEnter={() => setFidTooltipVisible(true)}
          onMouseLeave={() => setFidTooltipVisible(false)}>
          <ToolTip isVisible={fidTooltipVisible}>
            <b>First Input Delay.</b>
            <br />
            Not reported if the user never interacts with the page
            <br />
            {mapMetricRating(wvFid.metric_rating)}. {wvFid.value}ms
          </ToolTip>
          FID: <MetricRating rating={wvFid.metric_rating} />
        </div>
      )}
      {wvCls && (
        <div title="Cumulative Layout Shift"
          aria-label='CLS'
          onMouseEnter={() => setClsTooltipVisible(true)}
          onMouseLeave={() => setClsTooltipVisible(false)}>
          <ToolTip isVisible={clsTooltipVisible}>
            <b>Cumulative Layout Shift.</b>
            <br />
            {mapMetricRating(wvCls.metric_rating)}. {wvCls.metric_value?.toFixed(4)} ({wvCls.metric_delta?.toFixed(4)}). time: {wvCls.event_time?.toFixed(4)}
          </ToolTip>
          CLS: <MetricRating rating={wvCls.metric_rating} />
        </div>
      )}
      {wvTtfb && (
        <div title="Time to first byte"
          aria-label='TTFB'
          onMouseEnter={() => setTtfbTooltipVisible(true)}
          onMouseLeave={() => setTtfbTooltipVisible(false)}>
          <ToolTip isVisible={ttfbTooltipVisible}>
            <b>Time to first byte.</b>
            <br />
            {mapMetricRating(wvTtfb.metric_rating)}. {wvTtfb.value}ms
            <br />
            fetch: {wvTtfb.fetch_start?.toFixed(4)}<br />
            domain_lookup: {wvTtfb.domain_lookup_start?.toFixed(2)}<br />
            connect: {wvTtfb.connect_start?.toFixed(2)}<br />
            request:{wvTtfb.request_start?.toFixed(2)}<br />
            response: {wvTtfb.response_start?.toFixed(2)}<br />
            DOM Content loaded: {wvTtfb.domcontent_loaded_start?.toFixed(2)}<br />
            DOM complete:{wvTtfb.dom_complete?.toFixed(2)}<br />
            load start: {wvTtfb.load_start?.toFixed(2)}
          </ToolTip>
          TTFB: <MetricRating rating={wvTtfb.metric_rating} />
        </div>
      )}
      {wvFcp && (
        <div title="First Contentful Paint"
          aria-label='FCP'
          onMouseEnter={() => setFcpTooltipVisible(true)}
          onMouseLeave={() => setFcpTooltipVisible(false)}>
          <ToolTip isVisible={fcpTooltipVisible}>
            <b>First Contentful Paint</b>
            <br />
            {mapMetricRating(wvFcp.metric_rating)}. {wvFcp.value / 1000}s
          </ToolTip>
          FCP: <MetricRating rating={wvFcp.metric_rating} />
        </div>
      )}
    </ClsWrapper>
  )
}

export default WebVitals
