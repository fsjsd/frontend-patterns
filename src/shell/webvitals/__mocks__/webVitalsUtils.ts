import { MetricInfo, RegisterWebVitalsListenersProps } from "../webVitalsUtils";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const { registerWebVitalsListeners: discard, ...originalModule } = jest.requireActual('./webVitalsUtils');

export const registerWebVitalsListeners = ({
  onCls,
  onFcp,
  onLcp,
  onTtfb,
  onFid, }: RegisterWebVitalsListenersProps) => {
  onCls({ value: 1, metric_rating: "good" } as MetricInfo);
  onFcp({ value: 1, metric_rating: "poor" } as MetricInfo);
  onLcp({ value: 1, metric_rating: "poor" } as MetricInfo);
  onTtfb({ value: 1, metric_rating: "ni" } as MetricInfo);
  onFid({ value: 1, metric_rating: "" } as MetricInfo);
}

export default originalModule;