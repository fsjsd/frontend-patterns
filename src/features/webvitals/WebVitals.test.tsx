import React from 'react'
import { render } from '@testing-library/react'
import WebVitals from './WebVitals'
import { MetricInfo, RegisterWebVitalsListenersProps } from './webVitalsUtils';

jest.mock("./webVitalsUtils", () => {
  const originalModule = jest.requireActual('./webVitalsUtils');

  //Mock the default export and named export 'foo'
  return {
    __esModule: true,
    ...originalModule,
    registerWebVitalsListeners: ({
      onCls,
      onFcp,
      onLcp,
      onTtfb,
      onFid, }: RegisterWebVitalsListenersProps) => {
      onCls({ value: 1, metric_rating: "good" } as MetricInfo);
      onFcp({ value: 1, metric_rating: "poor" } as MetricInfo);
      onLcp({ value: 1, metric_rating: "good" } as MetricInfo);
      onTtfb({ value: 1, metric_rating: "good" } as MetricInfo);
      onFid({ value: 1, metric_rating: "good" } as MetricInfo);
    },
  }
});

describe('WebVitals', () => {
  test('render matches snapshot', () => {
    const { container } = render(<WebVitals />);
    expect(container).toMatchSnapshot()
  })
});