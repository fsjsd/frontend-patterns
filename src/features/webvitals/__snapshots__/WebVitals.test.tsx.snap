// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebVitals render matches snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ gUwqzG"
  >
    <div
      title="Largest Contentful Paint"
    >
      LCP:
      <div
        class="sc-gsnTZi eVhfZt"
      >
        good
      </div>
      (
      1
      )
    </div>
    <div
      title="First Input Delay. Not reported if the user never interacts with the page"
    >
      FID:
      <div
        class="sc-gsnTZi eVhfZt"
      >
        good
      </div>
    </div>
    <div
      title="Cumulative Layout Shift"
    >
      CLS:
      <div
        class="sc-gsnTZi eVhfZt"
      >
        good
      </div>
       
      (
      1
      )
    </div>
    <div
      title="Time to first byte"
    >
      TTFB:
      <div
        class="sc-gsnTZi eVhfZt"
      >
        good
      </div>
       
    </div>
    <div
      title="First Contentful Paint"
    >
      FCP:
      <div
        class="sc-gsnTZi bsTnoU"
      >
        poor
      </div>
       
    </div>
  </div>
</div>
`;
